src/error.rs

```diff
@@ -0,0 +1,52 @@
+use crate::middleware::{Statement, StatementTmpl};
+
+pub type Result<T, E = Error> = core::result::Result<T, E>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    // Wrappers on top of other errors
+    #[error("std::io::Error: {0}")]
+    IOError(#[from] std::io::Error),
+    #[error("anyhow::Error: {0}")]
+    Anyhow(#[from] anyhow::Error),
+    #[error(transparent)]
+    Infallible(#[from] std::convert::Infallible),
+
+    // sub-categories of errors
+    #[error(transparent)]
+    Tree(#[from] crate::backends::plonky2::primitives::error::TreeError),
+    #[error(transparent)]
+    Middleware(#[from] crate::middleware::error::MiddlewareError),
+
+    // Comparators errors
+    #[error("not equal")]
+    NotEqual,
+    #[error("{0} {1} is over the limit {2}")]
+    MaxLength(String, usize, usize),
+    #[error("{0} amount of {1} should be {1} but it's {2}")]
+    DiffAmount(String, String, usize, usize),
```

Comment:
Did you consider naming the fields so that the error creation is more clear?
For example:
```rust
DiffAmount{object: String, unit: String, expect: usize, found: usize}
```
The same applies to other error cases that have several fields